logstash-core/src/main/java/org/logstash/settings/SettingExt.java

```diff
@@ -0,0 +1,170 @@
+    @SuppressWarnings("unchecked")
+    public SettingExt initialize(ThreadContext context, IRubyObject[] args, final Block block) {
+        RubyString name = (RubyString) args[0];
+        RubyModule clazz = (RubyModule) args[1];
+        RubyObject defaultValue = null;
+        if (args.length >= 3) {
+            defaultValue = (RubyObject) args[2];
+        }
+        boolean strict = true;
+        if (args.length >= 4) {
+            strict = args[3].toJava(Boolean.class);
+        }
+
+        System.out.println("DNADBG >> creating setting for " + name.asJavaString());
+        final Class<?> javaClass = rubyClassToJava(clazz);
+        if (block.isGiven()) {
+            Predicate<Object> validator = new Predicate<Object>() {
+                @Override
+                public boolean test(Object o) {
+                    return block.yield(context, (IRubyObject) o).toJava(Boolean.class);
+                }
+            };
+
+            setting = new Setting(name.asJavaString(), javaClass, convertTo(defaultValue, javaClass), strict, validator);
+        } else {
+            final Object coercedValue = convertTo(defaultValue, javaClass);
+            setting = new Setting(name.asJavaString(), javaClass, coercedValue, strict);
+        }
+        return this;
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private Object convertTo(RubyObject defaultValue, Class<?> javaClass) {
+        if (defaultValue instanceof RubyArray) {
+            final RubyArray castedValue = (RubyArray) defaultValue;
+            return new ArrayList(castedValue);
+        }
+        if (TimeValue.class.equals(javaClass)) {
+            return TimeValue.fromValue(defaultValue.asJavaString());
+        }
+        return defaultValue.toJava(javaClass);
+    }
+
+    private Class<?> rubyClassToJava(RubyModule rclass) {
+        if ("String".equals(rclass.getName())) {
+            return String.class;
+        }
+        if ("Array".equals(rclass.getName())) {
+            return ArrayList.class;
+        }
+        if ("Java::OrgLogstashUtil::ModulesSettingArray".equals(rclass.getName())) {
+            return ModulesSettingArray.class;
+        }
+        if ("Java::OrgLogstashUtil::CloudSettingId".equals(rclass.getName())) {
+            return CloudSettingId.class;
+        }
+        if ("Java::OrgLogstashUtil::CloudSettingAuth".equals(rclass.getName())) {
+            return CloudSettingAuth.class;
+        }
+        if ("Java::OrgLogstashUtil::TimeValue".equals(rclass.getName())) {
+            return TimeValue.class;
+        }
+        // this cover Boolean (Ruby doesn't have a class for it) and StringCoercible
+        if ("Object".equals(rclass.getName())) {
+            return Object.class;
+        }
+        if ("Integer".equals(rclass.getName())) {
+            return Integer.class;
+        }
+        throw new IllegalArgumentException("Cannot find matching Java class for: " + rclass.getName());
```

Comment:
could this be pulled up into the specific `SettingOption` implementation referenced above? I'd like to avoid an ever-growing if chain. If we do leave the implementation here, perhaps a `switch` statement would be more efficient and read more clearly?

```suggestion
        switch (rclass.getName()) {
            case "String":
                return String.class;
            case "Array":
                return ArrayList.class;
            case "Java::OrgLogstashUtil::ModulesSettingArray":
                return org.logstash.util.ModulesSettingArray.class;
            case "Java::OrgLogstashUtil::CloudSettingId":
                return org.logstash.util.CloudSettingId.class;
            case "Java::OrgLogstashUtil::CloudSettingAuth":
                return org.logstash.util.CloudSettingAuth.class;
            case "Java::OrgLogstashUtil::TimeValue":
                return org.logstash.util.TimeValue.class;
            case "Object":
                return Object.class;
            case "Integer":
                return Integer.class;
            default:
                throw new IllegalArgumentException("Cannot find matching Java class for: " + rclass.getName());
        }
```

Reply:
Moved to switch, thanks.